components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -241,8 +241,6 @@ Note that if there is association between a federated user and local user () 'As
             if (ArrayUtils.contains(allowedResourcesForUser, ROOT) || ArrayUtils.contains(allowedResourcesForUser,
```

Comment:
Can we think of unifying the logic by rewriting this log to,

1. Have a method that evaluates requested scopes against user's permissions.
2. If a user is requesting SYSTEM scopes he is requesting all scopes in the system. We can use method introduced step 1 by passing all scopes in the system as request scopes.

Basic idea is we should not treat any user special and have short circuited paths like this :)

Reply:
If we assign `allScopes` to `requestedScopes` when the requested scope is SYSTEM, we will have to repeat 3 for loops that occure in line 246, 252, 256 again because `requestedScopes` is an array of strings but `allScopes` is a set of scopes.

Reply:
To do this, we have to introduce another for loop that iterates over all scopes as `requestedScopes` is an array of string but `allScopes` is a set of scopes. And it will decrease the efficiency as some other places (line 109, 154) will start to iterate over all the scopes if we didn't avoid for "SYSTEM" scope.

Reply:
requestedScopes --> can't we get this as a set?